### PR TITLE
feat: add EFP (Ethereum Follow Protocol) integration for wallet

### PR DESCRIPTION
**Pull request by caveman-eth**
Adds support for fetching and managing Ethereum Follow Protocol (EFP) followings in the wallet service

Relating to: https://github.com/status-im/status-desktop/issues/18686
and status-desktop PR: https://github.com/status-im/status-desktop/pull/19195

**Screenshot:**
efp

**Complete video demo:** 
https://streamable.com/3y4s1f

**Changes:**

**New EFP Client (`services/wallet/thirdparty/efp/`):**
- HTTP client for api.ethfollow.xyz
- Pagination support (limit/offset parameters)
- Server-side search via `/searchFollowing` endpoint
- Stats endpoint for total following count
- ENS name and avatar resolution included in responses

**Following Manager (`services/wallet/following/`):**
- Provider coordination layer
- Support for multiple data providers (extensible)
- Error handling and logging

**RPC API (`services/wallet/api.go`):**
- `wallet_GetFollowingAddresses(userAddress, search, limit, offset)` - fetch following addresses with optional search and pagination
- `wallet_GetFollowingStats(userAddress)` - get total following count

**Tests:**
- `client_test.go` - 6 unit tests for HTTP client (pagination, search, stats, error handling)
- `manager_test.go` - 5 unit tests for manager (success cases, provider errors, edge cases)
- Mock generation via gomock for testing

#### API Endpoints Used
- `GET /users/{address}/following?include=ens&limit={limit}&offset={offset}&sort=latest`
- `GET /users/{address}/searchFollowing?include=ens&sort=latest&term={search}`
- `GET /users/{address}/stats`

---

### CODE REVIEW

**Comment by caveman-eth** [comment]
@Khushboo-dev-cpp @dlipicar @jrainville :)

**Review by jrainville — APPROVED** [review]
Awesome work! Looks very good. 

You'll have to rebase and fix some lint issues I think

**Comment by jrainville** [comment]
ah, also commits need to follow conventional commit format, eg `feat: add EFP following address support`

**Comment by igor-sirotin** [comment]
@caveman-eth I have sent you an Github invite for collaborator. Please accept and rebase to trigger the CI 🙂

**Comment by caveman-eth** [comment]
> @caveman-eth I have sent you an Github invite for collaborator. Please accept and rebase to trigger the CI 🙂

Great, looks like everything is passing now

> Awesome work! Looks very good.
> 
> You'll have to rebase and fix some lint issues I think

@igor-sirotin I think everything is right now :)

File: services/wallet/following/manager.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/status-im/status-go/logutils"
+	"github.com/status-im/status-go/services/wallet/thirdparty/efp"
+)
+
+// Manager handles following address operations using EFP providers
+type Manager struct {
+	providers []efp.FollowingDataProvider
```
Comment (dlipicar) [inline services/wallet/following/manager.go]: we use a slice of providers for cases where multiple providers can be used to fetch the same type of data. I'm not super well versed in EFP, but it doesn't look like that's something we expect to happen, right?
Moreover, we're just using `m.providers[0]` in all methods, so making this a slice is misleading, better just make it `provider efp.FollowingDataProvider`

File: services/wallet/following/manager.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	logutils.ZapLogger().Debug("following.Manager.FetchFollowingAddresses",
+		zap.String("userAddress", userAddress.Hex()),
+		zap.String("search", search),
+		zap.Int("limit", limit),
+		zap.Int("offset", offset),
+		zap.Int("providers.len", len(m.providers)),
+	)
+
+	if len(m.providers) == 0 {
+		return []efp.FollowingAddress{}, nil
```
Comment (dlipicar) [inline services/wallet/following/manager.go]: feels like we should return an error if the provider is not set and we make a request, otherwise we can't distinguish this from a "true" empty FollowingAddresses result.

File: services/wallet/service.go [hunk 3]
```diff
@@ -246,6 +248,13 @@ func NewService(
 		collectiblesOwnershipController,
 		collectiblesPublisher)
 
+	// EFP (Ethereum Follow Protocol) providers
+	efpClient := efp.NewClient()
+	followingProviders := []efp.FollowingDataProvider{
```
Comment (dlipicar) [inline services/wallet/service.go]: Check how we conditionally populate the other provider lists https://github.com/status-im/status-go/blob/develop/services/wallet/service.go#L119
Please do the same thing here, we don't want accesses to the provider if `thirdpartyServicesEnabled` is false

File: services/wallet/thirdparty/efp/client.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+type Client struct {
+	httpClient *thirdparty.HTTPClient
+	baseURL    string
+}
+
+func NewClient() *Client {
+	httpClient := thirdparty.NewHTTPClient(
+		thirdparty.WithDetailedTimeouts(
+			5*time.Second,  // dialTimeout
```
Comment (dlipicar) [inline services/wallet/thirdparty/efp/client.go]: these seem to be opinionated default values without any way of overriding them. Perhaps pass the `httpClient` as a constructor parameter instead?

File: services/wallet/thirdparty/efp/client.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/status-im/status-go/services/wallet/thirdparty"
+)
+
+const baseURL = "https://api.ethfollow.xyz/api/v1"
+
+const (
+	requestDelay = 100 * time.Millisecond
```
Comment (dlipicar) [inline services/wallet/thirdparty/efp/client.go]: Is this used anywhere?

**Review by dlipicar — COMMENTED** [review]
Thanks for your contribution! A couple minor remarks added

File: services/wallet/following/manager.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+func NewManager(providers []efp.FollowingDataProvider) *Manager {
+	return &Manager{
+		providers: providers,
+	}
+}
+
+// FetchFollowingAddresses fetches the list of addresses that the given user is following
+// Uses the first available provider (can be enhanced later with fallback logic)
+func (m *Manager) FetchFollowingAddresses(ctx context.Context, userAddress common.Address, search string, limit, offset int) ([]efp.FollowingAddress, error) {
+	logutils.ZapLogger().Debug("following.Manager.FetchFollowingAddresses",
```
Comment (igor-sirotin) [inline services/wallet/following/manager.go]: Would be nice to pass a `logger` to `NewManager`. This way we'll be able to utilize logging namespaces.

Then use it like this everywhere:
```suggestion
	m.logger.Debug("following.Manager.FetchFollowingAddresses",
```

File: services/wallet/following/manager_test.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+	"github.com/status-im/status-go/services/wallet/thirdparty/efp"
+	mock_efp "github.com/status-im/status-go/services/wallet/thirdparty/efp/mock"
+)
+
+func TestFetchFollowingAddressesSuccess(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	ctx := context.TODO()
```
Comment (igor-sirotin) [inline services/wallet/following/manager_test.go]:
```suggestion
	ctx := t.Context()
```

Comment (igor-sirotin) [reply services/wallet/following/manager_test.go]: (and in other tests below)

File: services/wallet/thirdparty/efp/client_test.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	srv, stop := setupTest(t, response)
+	defer stop()
+
+	efpClient := &Client{
+		httpClient: thirdparty.NewHTTPClient(),
+		baseURL:    srv.URL,
+	}
+
+	userAddress := common.HexToAddress("0x742d35cc6cf4c7c7")
+	addresses, err := efpClient.FetchFollowingAddresses(context.Background(), userAddress, "", 10, 0)
```
Comment (igor-sirotin) [inline services/wallet/thirdparty/efp/client_test.go]:
```suggestion
	addresses, err := efpClient.FetchFollowingAddresses(t.Context(), userAddress, "", 10, 0)
```

File: services/wallet/thirdparty/efp/client_test.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+				Data:       "0x983110309620D911731Ac0932219af06091b6744",
+				Tags:       []string{"ens"},
+				ENS: &ENSData{
+					Name: "vitalik.eth",
+				},
+			},
+		},
+	}
+
+	response, _ := json.Marshal(expected)
```
Comment (igor-sirotin) [inline services/wallet/thirdparty/efp/client_test.go]:
```suggestion
	response, err := json.Marshal(expected)
	require.NoError(err)
```

File: services/wallet/thirdparty/efp/client_test.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		_, err := w.Write(response)
+		if err != nil {
+			return
+		}
+	}))
+
+	return srv, func() {
+		srv.Close()
+	}
+}
```
Comment (igor-sirotin) [inline services/wallet/thirdparty/efp/client_test.go:28]: Maybe implement a test suite, not to repeat this call in every test?
```go
type EFPClientTestSuite struct {
    suite.Suite
}

func (s *EFPClientTestSuite) SetupTest() {
     // setup
}

func (s *EFPClientTestSuite) TeardownTest() {
    // cleanup
}
```

**Review by igor-sirotin — APPROVED** [review]
I agree with comments from @dlipicar, otherwise looks good! Thanks for your contribution!
I've left some minor suggestions, no blockers from my side.

**Comment by caveman-eth** [comment]
@dlipicar and @igor-sirotin thanks for reviewing.
@dlipicar I have now implemented all of your suggestions, thanks for pointing out.
@igor-sirotin I have added all of your suggestions, aside from adding test suite. I could still do it, but didn't put it as high priority.

**Review by igor-sirotin — APPROVED** [review]
Thanks for the fixes 👍

**Review by dlipicar — APPROVED** [review]

**Comment by caveman-eth** [comment]
I have updated the branch with latest upstream commits, and resolved a status-go conflict.

**Comment by igor-sirotin** [comment]
@caveman-eth let's make sure https://github.com/status-im/status-desktop/pull/19195 is green and approved, I'll merge this one then
